### ISSUE TITLE
Memory leak and unessessary copy

### ISSUE THREAD

**Issue by LagradOst** [tirithen/clipper2#3]
For example `impl From<Polygons> for PolygonsC` 
1. Clones the entire content of Polygon instead of transmuting the pointer to the vec :-1:
2. Never cleans up the memory :-1:

Both of these are solved if the for example Polygons owns the data and PolygonsC is only a c pointer reference into Polygons so we dont need to do any memory cloning, and this makes rust clean up the memory. This assumes that all the c types are private, as PolygonsC must never outlive Polygons if this is the case.

**Comment by tirithen** [comment]
Thanks for the feedback! I know about the memory leaks, but only had time so far to get the ffi and build working.

My first idea was to copy the clipper-sys crate code, and using specific functions like free_path_c and so on to clean up the memory.

Your approach sounds much better if that makes the memory clean up automatically.

The clipper-sys crate exposed unsafe functions (it was made for use inside the geo-clipper crate), but my intention is to keep the unsafe functions internal and expose safe functions keeping the cleanup internally.

I initially now exposed the unsafe functions and types as well, but it is really not that useful if the C-types points to the same memory that seem much better to me as well, I'm all for reducing copies of memory.

I'll give it a go and see if I can get that to work.

**Comment by tirithen** [comment]
I have started on a refactor that will include reduction of copying, data transformation steps, and that also will clean up allocated memory.

https://github.com/tirithen/clipper2/pull/4

**Comment by tirithen** [comment]
@LagradOst I have been refactoring the code to provide a cleaner API and reducing the memory copy. Progress is a bit slow as I know much more about Rust than C++ and about how to work with bindgen.

It seems that the new [`Paths`](https://github.com/tirithen/clipper2/blob/no-copy/src/clipper2/path.rs#L3) type still does not use quite the correct memory layout as I get runtime errors when calling [`Clipper2Lib::Union`](https://github.com/tirithen/clipper2/blob/no-copy/clipper2/wrapper.cpp#L14).

The API is now as seen in https://github.com/tirithen/clipper2/blob/no-copy/src/tests.rs :
```rust
// shortest to type, but with extra conversion
let shape1 = vec![(1.0, 1.0), (1.0, 5.0), (5.0, 5.0), (5.0, 1.0)];
let shape2 = vec![(2.0, 2.0), (2.0, 4.0), (8.0, 4.0), (8.0, 2.0)];
let shapes = vec![shape1, shape2];

let paths = Paths::from(shapes);
let result = union(&paths, FillRule::NonZero);

// or the slightly more verbose version, but with fewer converisons
let shape1 = vec![Point::new(1.0, 1.0), Point::new(1.0, 5.0), Point::new(5.0, 5.0), Point::new(5.0, 1.0)];
let shape2 = vec![Point::new(2.0, 2.0), Point::new(2.0, 4.0), Point::new(8.0, 4.0), Point::new(8.0, 2.0)];
let shapes = vec![shape1, shape2];

let paths = Paths::from_vec(shapes);
let result = union(&paths, FillRule::NonZero);
```

In case it would interest you and you have some feedback on the API and possibly even some insights on how to work with the `Paths` like type across Rust/C++ while keeping memory copy to a minimum those things would be greatly appreciated.

I think this refactor now looks quite promising once the Rust/C++ interop errors can be solved.

**Comment by LagradOst** [comment]
After looking the code over, I can only say undefined behavior.
1. The class is not repr C, as rust does not have a defined struct layout you cant just cast it 😔 
https://github.com/tirithen/clipper2/blob/no-copy/src/clipper2/path.rs#L3
https://github.com/tirithen/clipper2/blob/no-copy/src/boolean.rs#L24

2. You cast to a vec with Vec::from_raw_parts and define your own capacity (+1), while I think this works it should be undefined behavior as the capacity is "lost"
https://github.com/tirithen/clipper2/blob/no-copy/src/clipper2/path.rs#L146

This makes so the code may work but shouldn't. You might actually want to use a const generic bool to define who owns the data (C or Rust). 

On the API Front:

Also if you already flatten the data here https://github.com/tirithen/clipper2/blob/no-copy/src/clipper2/path.rs#L11C12-L11C20 then you can just use impl Iterator and extend

**Comment by tirithen** [comment]
Thanks for looking on the code, I really appreciate it! As you probably can tell, the FFI parts with shared memory layout between Rust and C++ is a bit on the limits of my current ability. But I hope learn and get help to get past that as I really have great use cases for the Clipper2 library and I believe others also would benefit from it.

In the worst case I suppose I could revert back to serializing/copying the data as that should be simpler, but the shared memory should be much more efficient so I really want to give that a go.

You are definitely right that casting the struct without having it as `#[repr(C)]` does not make much sense as Rust gives no guarantees for the ordering in the memory layout. Thanks for pointing that out. I still thinking that casting should be okay IF I can get the ordering and sizes of the fields to stay the same in C++ and Rust.

As you also pointed out Paths would need to retain information both about the size and capacity of each std::vector as separate fields for the memory to map correctly, right now that information is not retained.  Possibly I should create a small separate project just to play with passing a single layer `Vec<Point>` in both directions (I'll need a `Vec<Vec<Point>>` in the end). That might be easier to understand for me.

As for the [`Paths::from_vec`](https://github.com/tirithen/clipper2/blob/no-copy/src/clipper2/path.rs#L11C12-L11C20) I take it that you mean I could use a signature something in the style of `pub fn from(data: Iterator<Item=Iterator<Item=Point>>)`? Rather than requiring it to be a `Vec<Vec<Point>>`? Even better if the Point part also was flexible as long as it is static dispatch (faster? and better for libraries as I have understood it).

Still learning a lot here, anyhow thank you a lot for the help so far. Hopefully I can soon find a way through this, I think I can see the possibility, just not clearly yet the steps needed to get there. :-)

**Comment by tirithen** [comment]
Oh, and by the way, any experience with https://cxx.rs/binding/cxxvector.html ? I'm wondering if that would simplify things, or just add another complexity layer.

Maybe it would just lead to more copying of data:

```rust

#![no_main] // main defined in C++ by main.cc

use cxx::{CxxString, CxxVector};

#[cxx::bridge]
mod ffi {
    extern "Rust" {
        fn f(vec: &CxxVector<CxxString>);
    }
}

fn f(vec: &CxxVector<CxxString>) {
    let vec: Vec<String> = vec
        .iter()
        .map(|s| s.to_string_lossy().into_owned())
        .collect();
    g(&vec);
}

fn g(vec: &[String]) {
    println!("{:?}", vec);
}
```

**Comment by tirithen** [comment]
@LagradOst I have now replaced the FFI code and the API. I was not able to get rid of the copying as I initially was hoping for, but I think that I got rid of the memory leaks. Changes has been released as version 0.2.0.

I think I would need more help to implement a zero copy version of this, and I think that is a separate issue from the memory leaks fixes so we should anyhow be better of with the new version.

I'm closing this issue now. But feel free to open another one with suggestions for any zero copy improvements that you can come up with. :-)
